test/parallel_api/numeric/numeric.ops/reduce_by_segment.pass.cpp

```diff
@@ -21,210 +21,208 @@
     {
-        auto keys    = key_buf.template get_access<sycl::access::mode::read_write>();
-        auto vals    = val_buf.template get_access<sycl::access::mode::read_write>();
-        auto keys_res    = key_res_buf.template get_access<sycl::access::mode::read_write>();
-        auto vals_res    = val_res_buf.template get_access<sycl::access::mode::read_write>();
-
-        //T keys[n1] = { 1, 2, 3, 4, 1, 1, 3, 3, 1, 1, 3, 3, 0 };
-        //T vals[n1] = { 1, 2, 3, 4, 1, 1, 3, 3, 1, 1, 3, 3, 0 };
-
-        // keys_result = {1, 2, 3, 4, 1, 3, 1, 3, 0};
-        // vals_result = {1, 2, 3, 4, 2, 6, 2, 6, 0};
-
-        // Initialize data
-        for (int i = 0; i != 12; ++i) {
-            keys[i] = i % 4 + 1;
-            vals[i] = i % 4 + 1;
-            keys_res[i] = 9;
-            vals_res[i] = 1;
-            if (i > 3) {
-                ++i;
-                keys[i] = keys[i-1];
-                vals[i] = vals[i-1];
-                keys_res[i] = 9;
-                vals_res[i] = 1;
-            }
+        //T keys[n1] = { 1, 2, 3, 4, 1, 1, 3, 3, 1, 1, 3, 3, ..., 0 };
+        //T vals[n1] = { 1, 2, 3, 4, 1, 1, 3, 3, 1, 1, 3, 3, ..., 0 };
+        for (int i = 0; i < n - 1; ++i)
+        {
+            std::size_t fill_first = i + 1;
+            std::size_t fill_repeat = i % 4 < 2 ? 1 : 3;
+            std::size_t fill = i < 4 ? fill_first : fill_repeat;
+
+            host_keys[i] = fill;
+            host_vals[i] = fill;
+            host_key_res[i] = 9;
+            host_val_res[i] = 1;
         }
-        keys[12] = 0;
-        vals[12] = 0;
+        host_keys[n - 1] = 0;
+        host_vals[n - 1] = 0;
     }
 
-    // create sycl iterators
-    auto key_beg = oneapi::dpl::begin(key_buf);
-    auto key_end = oneapi::dpl::end(key_buf);
-    auto val_beg = oneapi::dpl::begin(val_buf);
-    auto key_res_beg = oneapi::dpl::begin(key_res_buf);
-    auto val_res_beg = oneapi::dpl::begin(val_res_buf);
+    template <typename Iterator1, typename Iterator2, typename Size>
+    void
+    check_values(Iterator1 key_res, Iterator2 val_res, Size n)
+    {
+        // keys_result = {1, 2, 3, 4, 1, 3, 1, 3, ..., 0};
+        // vals_result = {1, 2, 3, 4, 2, 6, 2, 6, ..., 0};
+
+        for (auto i = 0; i != n; ++i)
+        {
+            if (i == n - 1)
+            {
+                EXPECT_TRUE(key_res[i] == 0, "wrong effect from reduce_by_segment");
+                EXPECT_TRUE(val_res[i] == 0, "wrong effect from reduce_by_segment");
+            }
+            else if (i < 4)
+            {
+                EXPECT_TRUE(key_res[i] == i + 1, "wrong effect from reduce_by_segment");
+                EXPECT_TRUE(val_res[i] == i + 1, "wrong effect from reduce_by_segment");
+            }
+            else if (i % 2 == 0)
+            {
+                EXPECT_TRUE(key_res[i] == 1, "wrong effect from reduce_by_segment");
+                EXPECT_TRUE(val_res[i] == 2 || val_res[i] == 1, "wrong effect from reduce_by_segment");
```

Comment:
As far as I understand, ```val_res[i] == 1``` may occur when the key is followed by 0 and does not have a pair, i.e. if ```i == n - 2```. If so, let's strengthen the check: ```... || (val_res[i] == 1 && i == n - 2)```

Reply:
Done.